engine/server/session.go

```diff
@@ -0,0 +1,953 @@
+
+	sess.stateMu.Lock()
+	defer sess.stateMu.Unlock()
+	switch sess.state {
+	case sessionStateUninitialized:
+		if err := srv.initializeDaggerSession(opts.ClientMetadata, sess, failureCleanups); err != nil {
+			return nil, nil, fmt.Errorf("initialize session: %w", err)
+		}
+	case sessionStateInitialized:
+		// nothing to do
+	case sessionStateRemoved:
+		return nil, nil, fmt.Errorf("session %q removed", sess.sessionID)
+	}
+
+	// get or initialize the client itself
+
+	sess.clientMu.Lock()
+	client, clientExists := sess.clients[clientID]
+	if !clientExists {
+		client = &daggerClient{
+			daggerSession: sess,
+			clientID:      clientID,
+			secretToken:   token,
+		}
+		sess.clients[clientID] = client
+
+		failureCleanups.Add("delete client ID", func() error {
+			sess.clientMu.Lock()
+			delete(sess.clients, clientID)
+			sess.clientMu.Unlock()
+			return nil
+		})
+	}
+	sess.clientMu.Unlock()
+
+	client.stateMu.Lock()
+	defer client.stateMu.Unlock()
+	switch client.state {
+	case clientStateUninitialized:
+		if err := srv.initializeDaggerClient(ctx, client, failureCleanups, opts); err != nil {
+			return nil, nil, fmt.Errorf("initialize client: %w", err)
+		}
+	case clientStateInitialized:
+		// verify token matches existing client
+		if token != client.secretToken {
+			return nil, nil, fmt.Errorf("client %q already exists with different secret token", clientID)
+		}
+	}
+
+	// increment the number of active connections from this client
+	client.activeCount++
+
+	return client, func() error {
+		client.stateMu.Lock()
+		defer client.stateMu.Unlock()
+		client.activeCount--
+
+		// if the main client caller has no more active calls, cleanup the whole session
+		if clientID != sess.mainClientCallerID {
+			return nil
+		}
+		if client.activeCount > 0 {
+			return nil
+		}
+
+		sess.stateMu.Lock()
+		defer sess.stateMu.Unlock()
+		switch sess.state {
+		case sessionStateInitialized:
+			return srv.removeDaggerSession(ctx, sess)
+		default:
+			// this should never happen unless there's a bug
+			slog.Error("session state being removed not in initialized state",
+				"session", sess.sessionID,
+				"state", sess.state,
+			)
+			return nil
+		}
+	}, nil
+}
+
+// ServeHTTP serves clients directly hitting the engine API (i.e. main client callers, not nested execs like module functions)
+func (srv *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	clientMetadata, err := engine.ClientMetadataFromHTTPHeaders(r.Header)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get client metadata: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if err := engine.CheckVersionCompatibility(clientMetadata.ClientVersion, engine.MinimumClientVersion); err != nil {
+		http.Error(w, fmt.Sprintf("incompatible client version: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	httpHandlerFunc(srv.serveHTTPToClient, &ClientInitOpts{
+		ClientMetadata: clientMetadata,
+	}).ServeHTTP(w, r)
+}
+
+// TODO: doc
```

Comment:
(just noting in case you wanted to get this in, nonblocking ofc)

Reply:
Doh on all accounts, fixed them all up